File: winograd.py
'''
author: niejiadong
date: 2020/04/22
'''
'''
x: (n, n)
w: (k, k)
'''
import numpy as np
import time

def timing(f):
    def inner(x, w):
        start = time.time()
        time.sleep(0.12)
        y = f(x, w)
        end = time.time()
        print('time：%s' % (end - start))
        return y

    return inner


@timing
def normal_conv2d(x, w):
    k, _ = w.shape
    n, _ = x.shape
    m = n - k + 1

    y = np.zeros((m, m))
    for i in range(m):
        for j in range(m):
            for p in range(k):
                for q in range(k):
                    y[i][j] += x[i+p][j+q] * w[p][q]
    
    return y


@timing
def im2col_conv2d(x, w):
    k, _ = w.shape
    n, _ = x.shape
    m = n - k + 1

    m2 = m*m
    k2 = k*k

    im2col = np.zeros((m2, k2))
    w = np.reshape(w, (k2,))

    for i in range(m):
        for j in range(m):
            for p in range(k):
                for q in range(k):
                    im2col[i*m+j][p*k+q] = x[i+p][j+q]
    
    y = im2col * w
    y = np.sum(y, axis=1)
    
    y = np.reshape(y, (m, m))
    return y


@timing
def winograd_f2_3_conv2d(x, w):
    k, _ = w.shape
    n, _ = x.shape
    m = n - k + 1
    assert(k == 3)
    assert(n % 2 == 0)
    assert(n >= 4)

    hf_m = int(m // 2)
    
    K0 = x[0:-2:2, :].reshape((hf_m, 1, n))
    K1 = x[1:-2:2, :].reshape((hf_m, 1, n))
    K2 = x[2::2, :].reshape((hf_m, 1, n))
    K3 = x[3::2, :].reshape((hf_m, 1, n))

    x = np.concatenate([K0 - K2, K1 + K2, K2 - K1, K1 - K3], axis=1)

    K0 = x[:, :, 0:-2:2].reshape((hf_m, 4, hf_m, 1))
    K1 = x[:, :, 1:-2:2].reshape((hf_m, 4, hf_m, 1))
    K2 = x[:, :, 2::2].reshape((hf_m, 4, hf_m, 1))
    K3 = x[:, :, 3::2].reshape((hf_m, 4, hf_m, 1))

    x = np.concatenate([K0 - K2, K1 + K2, K2 - K1, K1 - K3], axis=3)
    x = np.swapaxes(x, 1, 2)

    w = np.stack([w[0, :], (w[0, :] + w[1, :] + w[2, :]) / 2, (w[0, :] - w[1, :] + w[2, :]) / 2, w[2, :]], axis=0)
    w = np.stack([w[:, 0], (w[:, 0] + w[:, 1] + w[:, 2]) / 2, (w[:, 0] - w[:, 1] + w[:, 2]) / 2, w[:, 2]], axis=1)
    
    y = x * w  # 对于(4,4) vs (3,3)的卷积，原始要36次乘法，这里只要16次乘法
    y = np.stack((y[:,:,:,0] + y[:,:,:,1] + y[:,:,:,2], y[:,:,:,1] - y[:,:,:,2] - y[:,:,:,3]), axis=-1)
    y = np.stack((y[:,:,0,:] + y[:,:,1,:] + y[:,:,2,:], y[:,:,1,:] - y[:,:,2,:] - y[:,:,3,:]), axis=-1)

    y = y.reshape(hf_m, m, 2)
    y = np.swapaxes(y, 1, 2)
    y = y.reshape(m, m)
    
    return y


@timing
def winograd_f4_3_conv2d(x, w):
    k, _ = w.shape
    n, _ = x.shape
    m = n - k + 1
    assert(k == 3)
    assert((n - 6) % 4 == 0)
    assert(n >= 6)

    h_m = int((n - 6) / 4 + 1)  # int(m // 4)
    
    K0 = x[0:-2:4, :].reshape((h_m, 1, n))
    K1 = x[1:-2:4, :].reshape((h_m, 1, n))
    K2 = x[2::4, :].reshape((h_m, 1, n))
    K3 = x[3::4, :].reshape((h_m, 1, n))
    K4 = x[4::4, :].reshape((h_m, 1, n))
    K5 = x[5::4, :].reshape((h_m, 1, n))

    x = np.concatenate([
        4 * K0 - 5 * K2 + K4,
        -4 * K1 - 4 * K2 + K3 + K4,
        4 * K1 - 4 * K2 - K3 + K4,
        -2 * K1 - K2 + 2 * K3 + K4,
        2 * K1 - K2 - 2 * K3 + K4,
        4 * K1 - 5 * K3 + K5], axis=1)

    K0 = x[:, :, 0:-2:4]
    K1 = x[:, :, 1:-2:4]
    K2 = x[:, :, 2::4]
    K3 = x[:, :, 3::4]
    K4 = x[:, :, 4::4]
    K5 = x[:, :, 5::4]

    x = np.stack([
        4 * K0 - 5 * K2 + K4,
        -4 * K1 - 4 * K2 + K3 + K4,
        4 * K1 - 4 * K2 - K3 + K4,
        -2 * K1 - K2 + 2 * K3 + K4,
        2 * K1 - K2 - 2 * K3 + K4,
        4 * K1 - 5 * K3 + K5], axis=-1)
    x = np.swapaxes(x, 1, 2)

    w = np.stack([
        w[0, :] * 0.25,
        -(w[0, :] + w[1, :] + w[2, :]) / 6,
        (- w[0, :] + w[1, :] - w[2, :]) / 6,
        (w[0, :] + w[1, :] * 2 + w[2, :] * 4) / 24,
        (w[0, :] - w[1, :] * 2 + w[2, :] * 4) / 24,
        w[2, :]], axis=0)
    w = np.stack([
        w[:, 0] * 0.25,
        -(w[:, 0] + w[:, 1] + w[:, 2]) / 6,
        (- w[:, 0] + w[:, 1] - w[:, 2]) / 6,
        (w[:, 0] + w[:, 1] * 2 + w[:, 2] * 4) / 24,
        (w[:, 0] - w[:, 1] * 2 + w[:, 2] * 4) / 24,
        w[:, 2]], axis=1)
    
    y = x * w  # 对于(6,6) vs (3,3)的卷积，原始要144次乘法，这里只要36次乘法
    y = np.stack([
        y[:,:,:,0] + y[:,:,:,1] + y[:,:,:,2] + y[:,:,:,3] + y[:,:,:,4],
        y[:,:,:,1] - y[:,:,:,2] + y[:,:,:,3] * 2 - y[:,:,:,4] * 2,
        y[:,:,:,1] + y[:,:,:,2] + y[:,:,:,3] * 4 + y[:,:,:,4] * 4,
        y[:,:,:,1] - y[:,:,:,2] + y[:,:,:,3] * 8 - y[:,:,:,4] * 8 + y[:,:,:,5]
        ], axis=-1)
    y = np.stack([
        y[:,:,0,:] + y[:,:,1,:] + y[:,:,2,:] + y[:,:,3,:] + y[:,:,4,:],
        y[:,:,1,:] - y[:,:,2,:] + y[:,:,3,:] * 2 - y[:,:,4,:] * 2,
        y[:,:,1,:] + y[:,:,2,:] + y[:,:,3,:] * 4 + y[:,:,4,:] * 4,
        y[:,:,1,:] - y[:,:,2,:] + y[:,:,3,:] * 8 - y[:,:,4,:] * 8 + y[:,:,5,:]
        ], axis=-1)

    y = y.reshape(h_m, m, 4)
    y = np.swapaxes(y, 1, 2)
    y = y.reshape(m, m)
    
    return y


if __name__=='__main__':
    x = np.array(range(161604)).reshape((402, 402))
    # x = np.array(range(100)).reshape((10, 10))
    w = np.array(range(9)).reshape((3, 3))
    # w = np.ones((3,3))

    # y1 = normal_conv2d(x, w)
    # print(y1)

    # y2 = im2col_conv2d(x, w)
    # print(y2)

    y3 = winograd_f2_3_conv2d(x, w)
    print(y3)

    y4 = winograd_f4_3_conv2d(x, w)
    print(y4)

File: winograd.cpp
/*
author: niejiadong
date: 2020/04/22
*/
/*
x: (n, n)
w: (k, k)
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>


void normal_conv2d(float* x, float* w, float* y, int n)
{
    int k = 3;
    int m = n - 2;

    memset(y, 0, sizeof(float) * m * m);

    for (int i=0; i<m; i++) {
        for (int j=0; j<m; j++) {
            for (int p=0; p<3; p++) {
                for (int q=0; q<3; q++) {
                    y[i * m + j] += x[(i + p) * n + j + q] * w[p * k + q];
                }
            }
        }
    }
    
    return;
}


void im2col_conv2d(float* x, float* w, float* y, int n)
{
    int k = 3;
    int m = n - 2;
    int m2 = m * m;

    memset(y, 0, sizeof(float) * m2);

    float *im2col = (float*)malloc(m2 * 9 * sizeof(float));
    for (int i=0; i<m; i++) {
        for (int j=0; j<m; j++) {
            for (int p=0; p<3; p++) {
                for (int q=0; q<3; q++) {
                    im2col[(i * m + j) * 9 + p * k + q] = x[(i + p) * n + j + q];
                }
            }
        }
    }

    // can use gemm to accelerate, just multiply one by one now.
    for (int i=0; i<m2; i++) {
        for (int j=0; j<9; j++) {
            y[i] += im2col[i * 9 + j] * w[j];
        }
    }

    free(im2col);

    return;
}

void winograd_f2_3_conv2d(float* x, float* w, float* y, int n)
{
    int k = 3;
    int m = n - 2;
    int hf_m = m / 2;

    if (n < 4 && n % 2 != 0) {
        printf("Not support!");
        return;
    }
    
    float *B0 = (float*)malloc(n * hf_m * 4 * sizeof(float));
    float *B = (float*)malloc(hf_m * hf_m * 16 * sizeof(float));
    float *G = (float*)malloc(16 * sizeof(float));

    int i4;
    for (int i=0; i<3; i++) {
        i4 = i << 2;
        G[i4] = w[i * 3];
        G[i4 + 1] = (w[i * 3] + w[i * 3 + 1] + w[i * 3 + 2]) / 2;
        G[i4 + 2] = (w[i * 3] - w[i * 3 + 1] + w[i * 3 + 2]) / 2;
        G[i4 + 3] = w[i * 3 + 2];
    }

    for (int i=0; i<4; i++) {
        G[12 + i] = G[8 + i];
        G[8 + i] = (G[i] - G[4 + i] + G[12 + i]) / 2;
        G[4 + i] = (G[i] + G[4 + i] + G[12 + i]) / 2;
    }

    int offset, j4, j2, offset2;
    for (int i=0; i<n; i++) {
        offset = (i * hf_m) << 2;
        offset2 = i * n;
        for (int j=0; j<hf_m; j++) {
            j4 = (j << 2) + offset;
            j2 = (j << 1) + offset2;
            B0[j4] = x[j2] - x[j2 + 2];
            B0[j4 + 1] = x[j2 + 1] + x[j2 + 2];
            B0[j4 + 2] = x[j2 + 2] - x[j2 + 1];
            B0[j4 + 3] = x[j2 + 1] - x[j2 + 3];
        }
    }

    // 这一步花了比较长时间，内存不连续操作
    int j16, stride, i2;
    float *k0, *k1, *k2, *k3;
    for (int i=0; i<hf_m; i++) {
        offset = (i * hf_m) << 4;
        stride = hf_m << 2;
        offset2 = offset >> 1;
        k0 = B0 + offset2;
        k1 = k0 + stride;
        k2 = k1 + stride;
        k3 = k2 + stride;
        for (int j=0; j<hf_m; j++) {
            j16 = (j << 4) + offset;
            j4 = j << 2;
            for (int k=0; k<4; k++) {
                B[j16 + k] = k0[j4 + k] - k2[j4 + k];
                B[j16 + 4 + k] = k1[j4 + k] + k2[j4 + k];
                B[j16 + 8 + k] = k2[j4 + k] - k1[j4 + k];
                B[j16 + 12 + k] = k1[j4 + k] - k3[j4 + k];
            }
        }
    }
    
    int hf_m_2 = hf_m * hf_m;
    for (int i=0; i<hf_m_2; i++) {
        offset = i << 4;
        for (int j=0; j<16; j++) {
            B[offset + j] *= G[j];
        }
    }

    for (int i=0; i<hf_m_2; i++) {
        offset = i << 4;
        for (int j=0; j<4; j++) {
            j4 = (j << 2) + offset;
            B[j4] += B[j4 + 1] + B[j4 + 2];
            B[j4 + 1] -= B[j4 + 2] + B[j4 + 3];
        }
    }

    for (int i=0; i<hf_m_2; i++) {
        offset = i << 4;
        B[offset] += B[offset + 4] + B[offset + 8];
        B[offset + 4] -= B[offset + 8] + B[offset + 12];
        B[offset + 1] += B[offset + 5] + B[offset + 9];
        B[offset + 5] -= B[offset + 9] + B[offset + 13];
    }

    // reshape
    for (int i=0; i<hf_m; i++) {
        offset = (i << 1) * m;
        offset2 = (i * hf_m) << 4;
        for (int j=0; j<hf_m; j++) {
            y[offset + (j << 1)] = B[offset2 + (j << 4)];
            y[offset + (j << 1) + 1] = B[offset2 + (j << 4) + 1];
            y[offset + m + (j << 1)] = B[offset2 + (j << 4) + 4];
            y[offset + m + (j << 1) + 1] = B[offset2 + (j << 4) + 5];
        }
    }

    free(B0);
    free(B);
    free(G);

    return;
}


void print(float* y, int size)
{
    int max = 20;
    max = (max > size ? size : max);
    for (int i=0; i<max; i++) {
        printf("%.0f, ", y[i]);
    }
    printf("\n...\n");
    for (int i=max; i>0; i--) {
        printf("%.0f, ", y[size-i]);
    }
    printf("\n\n");
}


int main(void) {
    int n = 1000;
    int k = 3;
    int m = n - 2;
    float* x = (float*)malloc(sizeof(float) * n * n);
    float* w = (float*)malloc(sizeof(float) * k * k);
    float* y = (float*)malloc(sizeof(float) * m * m);

    for (int i=0; i<9; i++) {
        w[i] = i;
        // w[i] = 1;
    }
    for (int i=0; i<n*n; i++) {
        x[i] = i;
    }

    struct timeval tv1, tv2;

    gettimeofday(&tv1, NULL);
    normal_conv2d(x, w, y, n);
    gettimeofday(&tv2, NULL);
    printf("spend %ld us\n", (long)(tv2.tv_usec + 1000000 * tv2.tv_sec) - (tv1.tv_usec + 1000000 * tv1.tv_sec));
    print(y, m*m);

    // gettimeofday(&tv1, NULL);
    // im2col_conv2d(x, w, y, n);
    // gettimeofday(&tv2, NULL);
    // printf("spend %ld us\n", (long)(tv2.tv_usec + 1000000 * tv2.tv_sec) - (tv1.tv_usec + 1000000 * tv1.tv_sec));
    // print(y, m*m);

    gettimeofday(&tv1, NULL);
    winograd_f2_3_conv2d(x, w, y, n);
    gettimeofday(&tv2, NULL);
    printf("spend %ld us\n", (long)(tv2.tv_usec + 1000000 * tv2.tv_sec) - (tv1.tv_usec + 1000000 * tv1.tv_sec));
    print(y, m*m);

    free(x);
    free(w);
    free(y);

    return 0;
}